### ISSUE TITLE
No line number or filename for SyntaxErrors

### ISSUE THREAD

**Issue by fatcerberus** [svaarala/duktape#140]
Currently when Duktape throws a SyntaxError while compiling, there's no way to retrieve the filename or line number of the error as the error object placed on the stack doesn't contain that information and since it's a SyntaxError, there's no callstack to pull it from either.

Since I'm using Duktape in a game engine (minisphere) that uses JS for coding game logic and such games can comprise many JS files, not being able to see the exact location of the error is quite a nuisance.  Any possibility this could be fixed?

**Comment by svaarala** [comment]
Agree, there's room for improvement here. The basic error mechanics don't consider the code being compiled a part of the callstack so it doesn't work in the expected way, and I haven't had time to fix this yet.

But it's certainly something to fix.

**Comment by svaarala** [comment]
Right now the stack trace (recorded into the internal property _Tracedata) is:

```
SyntaxError: parse error (line 1)
	duk_js_compiler.c:3395
	eval  native strict directeval preventsyield
	global input:1 preventsyield
```

Getters inherited from Error.prototype are used to provide `fileName`, `lineNumber`, and `stack`.

In the current traceback the topmost entry (duk_js_compiler.c) is flagged as being "don't blame this for fileName/lineNumber" because it's not relevant for the user. This is why fileName and lineNumber end up undefined for SyntaxErrors.

One simple fix would be to somehow ensure that the function/line being compiled would be recorded into the _Tracedata similarly to how native functions are now recorded - but with the "blame" flag set.

Something like this (manipulating _Tracedata manually here):

```
duk> try { eval('foo=') } catch (e) { this.err = e; }
= SyntaxError: parse error (line 1)
duk> var tracedata = err[String(Duktape.dec('hex', 'ff')) + 'Tracedata']
= undefined
duk> tracedata
= duk_js_compiler.c,4294970691,function eval() {/* native */},107374182400,function global() {/* ecmascript */},34359738374
duk> tracedata[0] = 'my_source_file.js'
= my_source_file.js
duk> tracedata[1] = (0 * 4294967296.0 /*flags*/) + 1234 /*line*/;
= 1234
duk> err.stack
= SyntaxError: parse error (line 1)
	my_source_file.js:1234
	eval  native strict directeval preventsyield
	global input:1 preventsyield
duk> err.fileName
= my_source_file.js
duk> err.lineNumber
= 1234
```

**Comment by svaarala** [comment]
Here's a little test branch (not ideal but works):

- https://github.com/svaarala/duktape/compare/improve-syntax-error-fileline

Quick test:

```
((o) Duktape 1.1.99 (v1.1.0-194-g78278d8)
duk> try { eval('\n\nfoo='); } catch (e) { print(e.fileName, e.lineNumber); print(e.stack); }
input 3
SyntaxError: parse error (line 3)
	input:3
	eval  native strict directeval preventsyield
	global input:1 preventsyield
= undefined
```

And:

```
$ ./duk syntax_error.js 
SyntaxError: parse error (line 7)
	syntax_error.js:7
error in executing file syntax_error.js
```

This could be improved by keeping the duk_js_compiler.c in the callstack as well.

**Comment by fatcerberus** [comment]
I want to try this out in my game engine, but I don't know how to generate the monolithic duktape.c that I need.  Is there a script or something you use that does this?

**Comment by svaarala** [comment]
Cool - I'll give this some thought if there's a better fix. I'll merge some fix into master next week :)

**Comment by fatcerberus** [comment]
Interesting, this also apparently fixed the issue of no file and line for "invalid base value" errors, which were previously missing it.

**Comment by svaarala** [comment]
A better fix is now in master and will be part of 1.2.0 release.

**Comment by fatcerberus** [comment]
Nice.  This and the recent performance enhancements will make duktape even more awesome as the frontend of my minisphere game engine. :-D
